core/lib/libtomcrypt/src/ciphers/aes/aes_armv8a_ce.c

```diff
@@ -299,11 +299,17 @@ static int aes_ctr_encrypt_nblocks(const unsigned char *pt, unsigned char *ct,
 		return CRYPT_ERROR;
 	}
 
+	/* In big endian mode IV_size=0 -> 128 bits and IV_size=8 -> 64 bits */
+	if ((IV_size != 0) && (IV_size != 8)) {
+		/* Accelerated algorithm supports only 64 or 128 bits IV */
+		return CRYPT_ERROR;
+	}
+
 	Nr = skey->rijndael.Nr;
 	rk = (u8 *)skey->rijndael.eK;
 
 	tomcrypt_arm_neon_enable(&state);
-	ce_aes_ctr_encrypt(ct, pt, rk, Nr, blocks, IV, 1);
+	ce_aes_ctr_encrypt(ct, pt, rk, Nr, blocks, IV, IV_size, 1);
```

Comment:
You are passing `IV_size` for a parameter that is called `ctrlen`. I'm confused. Is the IV the same thing as the counter? (initial counter value?) The names should be consistent, and I'd like to have a detailed explanation of how the IV/counter value is formatted and incremented in the commit description.
A link to the CENC standard could be useful too, although I understand it is not freely available (ISO standard?). At least I'd like the precise reference of this standard, and a quick summary of what it says.
Without this I have no idea if what your are proposing is correct or not.
How did you test this code? Any new test case in `optee_test`?

Reply:
The general description of CTR mode of operation is described [here](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Counter_.28CTR.29).

AFAIU abut Libtomcrypt implementation of CTR mode, the buffer passed is an IV cum counter block (fixed block length: 16 bytes and combined together using addition operation). The variable length parameter we are talking about here is basically how much portion of this IV cum counter block is affected via an increment operation. Following are the scenarios we have to satisfy:
- CENC standard ([here](https://www.iso.org/obp/ui/#iso:std:iso-iec:23001:-7:ed-3:v1:en)): It allows for IV cum counter value where only 64 LSB bits are affected via counter increments like if we start with an IV cum counter value of 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFE, after we increment twice, we should get 0xFFFFFFFFFFFFFFFF0000000000000000.
- Default case (16 bytes of IV cum counter value): In this case all 128 bits are affected via counter increments like if we start with an IV cum counter value of 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFE, after we increment twice, we should get 0x00000000000000000000000000000000.

Hence we could say that the buffer isn't true IV or counter but its kind of combined form. That's why I have used `IV_size` where buffer is named as `IV` and `ctrlen` where buffer is named as `ctr` for consistency. However, I am still open to any other naming convention if that is able to make things more clearer.

And Libtomcrypt software implementation supports all value of IV cum counter length (1 - 16 bytes) but CE implementation was only limited to support 16 bytes IV cum counter value but this PR adds support for 8 bytes IV cum counter value to support CENC case.

> How did you test this code?

Just use following patch to enable 8 bytes IV cum counter value:

```
diff --git a/core/lib/libtomcrypt/ctr.c b/core/lib/libtomcrypt/ctr.c
index 1462f23..ed1fd1b 100644
--- a/core/lib/libtomcrypt/ctr.c
+++ b/core/lib/libtomcrypt/ctr.c
@@ -46,7 +46,7 @@ static TEE_Result ltc_ctr_init(struct crypto_cipher_ctx *ctx,
                c->update = ctr_decrypt;
 
        if (ctr_start(c->cipher_idx, iv, key1, key1_len, 0,
-                     CTR_COUNTER_BIG_ENDIAN, &c->state) == CRYPT_OK)
+                     CTR_COUNTER_BIG_ENDIAN | 8, &c->state) == CRYPT_OK)
                return TEE_SUCCESS;
        else
                return TEE_ERROR_BAD_STATE;
```

> Any new test case in optee_test?

I am not sure if we can make IV cum counter length as a configurable parameter to be passed from a TA. Maybe we can add a `CFG_*` compile time option for `optee_os`.

Reply:
@b49020 thanks, your explanation is quite clear :+1:

I checked the GP specs, and while v1.1 had "The counter MUST be encoded as a
16-byte buffer in big-endian form [...]", v1.2 now reads: "The counter SHALL be encoded as a
16-byte buffer in big-endian form [...]". Which in practice does not change a thing, because "SHALL indicates an absolute requirement, as does MUST"...